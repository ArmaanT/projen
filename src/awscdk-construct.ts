import { Component } from './component';
import { ConstructLibrary, ConstructLibraryOptions } from './construct-lib';

/**
 * Options for the construct-lib-aws project.
 */
export interface AwsCdkConstructLibraryOptions extends ConstructLibraryOptions {
  /**
   * Minimum target version this library is tested against.
   *
   * @default "1.78.0"
   */
  readonly cdkVersion: string;

  /**
   * Use pinned version instead of caret version for CDK.
   *
   * You can use this to prevent yarn to mix versions for your CDK dependencies and to prevent auto-updates.
   * If you use experimental features this will let you define the moment you include breaking changes.
   *
   * @default false
   */
  readonly cdkVersionPinning?: boolean;

  /**
   * Which AWS CDK modules (those that start with "@aws-cdk/") does this library
   * require when consumed?
   */
  readonly cdkDependencies?: string[];

  /**
   * Install the @aws-cdk/assert library?
   * @default true
   */
  readonly cdkAssert?: boolean;

  /**
   * AWS CDK modules required for testing.
   */
  readonly cdkTestDependencies?: string[];

  /**
   * Adds `integ:synth`, `integ:deploy`, and `integ:diff` tasks and a sample
   * integration test which can be used to validate this library.
   *
   * @default true
   */
  readonly cdkIntegTest?: boolean;
}

/**
 * AWS CDK construct library project
 *
 * A multi-language (jsii) construct library which vends constructs designed to
 * use within the AWS CDK with a friendly workflow and automatic publishing to
 * the construct catalog.
 *
 * ```ts
 * const project = new ConstructLibraryAws({
 *   name: 'cdk-watchful',
 *   description: 'Watching your CDK apps since 2019',
 *   jsiiVersion: Semver.caret('1.7.0'),
 *   authorName: 'Elad Ben-Israel',
 *   authorEmail: 'elad.benisrael@gmail.com',
 *   repository: 'https://github.com/eladb/cdk-watchful.git',
 *   keywords: [
 *     "cloudwatch",
 *     "monitoring"
 *   ],
 *
 *   catalog: {
 *     twitter: 'emeshbi'
 *   },
 *
 *   // creates PRs for projen upgrades
 *   projenUpgradeSecret: 'PROJEN_GITHUB_TOKEN',
 *
 *   cdkVersion: '1.54.0',
 *   cdkDependencies: [
 *     "@aws-cdk/aws-apigateway",
 *     "@aws-cdk/aws-cloudwatch",
 *     "@aws-cdk/aws-cloudwatch-actions",
 *     "@aws-cdk/aws-dynamodb",
 *     "@aws-cdk/aws-ecs",
 *     "@aws-cdk/aws-ecs-patterns",
 *     "@aws-cdk/aws-elasticloadbalancingv2",
 *     "@aws-cdk/aws-events",
 *     "@aws-cdk/aws-events-targets",
 *     "@aws-cdk/aws-lambda",
 *     "@aws-cdk/aws-rds",
 *     "@aws-cdk/aws-sns",
 *     "@aws-cdk/aws-sns-subscriptions",
 *     "@aws-cdk/aws-sqs",
 *     "@aws-cdk/core"
 *   ],
 *   devDependencies: {
 *     "aws-sdk": Semver.caret("2.708.0")
 *   },
 *
 *   // jsii publishing
 *
 *   java: {
 *     javaPackage: 'com.github.eladb.watchful',
 *     mavenGroupId: 'com.github.eladb',
 *     mavenArtifactId: 'cdk-watchful'
 *   },
 *   python: {
 *     distName: 'cdk-watchful',
 *     module: 'cdk_watchful'
 *   }
 * });
 *
 * project.synth();
 * ```
 *
 * @pjid awscdk-construct
 */
export class AwsCdkConstructLibrary extends ConstructLibrary {
  /**
   * The target CDK version for this library.
   */
  public readonly cdkVersion: string;

  constructor(options: AwsCdkConstructLibraryOptions) {
    super({
      ...options,
      peerDependencyOptions: {
        pinnedDevDependency: false,
      },
    });

    this.cdkVersion = options.cdkVersionPinning ? options.cdkVersion : `^${options.cdkVersion}`;

    this.addPeerDeps('constructs@^3.2.27');

    if (options.cdkAssert ?? true) {
      this.addDevDeps(this.formatModuleSpec('@aws-cdk/assert'));
    }

    this.addCdkDependencies(...options.cdkDependencies ?? []);
    this.addCdkTestDependencies(...options.cdkTestDependencies ?? []);
  }

  /**
   * Adds CDK modules as runtime dependencies.
   *
   * Modules are currently added with a caret CDK version both as "dependencies"
   * and "peerDependencies". This is because currently npm would not
   * automatically install peer dependencies that are not declared as concerete
   * dependencies by the consumer, so this is a little npm "hack" so that
   * consumers will not need to depend on them directly if they don't interact
   * with them.
   *
   * @param deps names of cdk modules (e.g. `@aws-cdk/aws-lambda`).
   */
  public addCdkDependencies(...deps: string[]) {
    // this ugliness will go away in cdk v2.0
    this.addPeerDeps(...deps.map(m => this.formatModuleSpec(m)));
    this.addDeps(...deps.map(m => this.formatModuleSpec(m)));
  }

  /**
   * Adds CDK modules as test dependencies.
   *
   * @param deps names of cdk modules (e.g. `@aws-cdk/aws-lambda`).
   */
  public addCdkTestDependencies(...deps: string[]) {
    this.addDevDeps(...deps.map(m => this.formatModuleSpec(m)));
  }

  private formatModuleSpec(module: string): string {
    return `${module}@${this.cdkVersion}`;
  }
}

interface IntegrationTestOptions {
  /**
   * The integration test app entrypoint.
   *
   * @default "test/integ.ts"
   */
  readonly entrypoint?: string;
}

class IntegrationTest extends Component {
  public readonly entrypoint: string;

  constructor(project: AwsCdkConstructLibrary, options: IntegrationTestOptions = {}) {
    super(project);

    this.entrypoint = options.entrypoint ?? 'test/integ.ts';

    // install the CLI and ts-node so we can use it to run the test
    project.addCdkTestDependencies('aws-cdk');
    project.addDevDeps('ts-node');

    const app = `ts-node --project ${project.tsconfigTest} ${this.entrypoint}`;
    const args = [
      '--app',
      `"${app}"`,
      '--context',
      '@aws-cdk/core:newStyleStackSynthesis=true',
    ];

    const synth = project.addTask('integ:synth', {
      description: 'Synthesize the integration test app to cdk.out',
    });

    project.gitignore.exclude('cdk.out');
    synth.exec('rm -fr cdk.out');
    synth.exec(`cdk synth ${args}`);

    project.addTask('integ:bootstrap', {
      description: 'Bootstrap AWS environments used by the integration test app',
      exec: `cdk bootstrap ${args}`,
      env: { CDK_NEW_BOOTSTRAP: '1' },
    });

    project.addTask('integ:deploy', {
      description: 'Deploy the integration test app',
      exec: `cdk deploy ${args} *`,
    });

    project.addTask('integ:diff', {
      description: 'Diff integration test app against AWS environment',
      exec: `cdk diff ${args}`,
    });

    project.testTask.spawn(synth);
  }
}

/** @deprecated use `AwsCdkConstructLibraryOptions` */
export interface ConstructLibraryAwsOptions extends AwsCdkConstructLibraryOptions { }

/** @deprecated use `AwsCdkConstructLibrary` */
export class ConstructLibraryAws extends AwsCdkConstructLibrary { }
