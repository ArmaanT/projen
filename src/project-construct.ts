import { Construct } from 'constructs';

export class ProjectConstruct extends Construct {
  /**
   * Synthesizes files to the project output directory.
   * @param _outdir The project directory*
   */
  public synthesize(_outdir: string) {}

  /**
   * Called after synthesis. Order is *not* guaranteed.
   * @param _outdir The project directory
   */
  public postSynthesize(_outdir: string) {}
}