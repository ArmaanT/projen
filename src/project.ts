import * as chalk from 'chalk';
import { Construct, ConstructNode } from 'constructs';

import { cleanup } from './cleanup';
import { printStartMenu } from './cli/cmds/start-app';
import { Component } from './component';
import { IgnoreFile } from './ignore-file';
import * as logging from './logging';
import { ProjectConstruct } from './project-construct';
import { SampleReadme } from './readme';
import { Start } from './start';

/**
 * Base project
 */
export class Project extends Construct {
  public readonly gitignore: IgnoreFile;
  public readonly name: string;

  private readonly components = new Array<Component>();
  private readonly tips = new Array<string>();

  constructor(name: string) {
    super(undefined as any, name);

    this.name = name;
    this.gitignore = new IgnoreFile(this, '.gitignore');

    new SampleReadme(this, '# my project');
  }

  /**
   * Prints a "tip" message during synthesis.
   * @param message The message
   */
  public addTip(message: string) {
    this.tips.push(message);
  }

  /**
   * Synthesize all project files into `outdir`
   * @param outdir The project root directory (default is `.`).
   */
  public synth(outdir: string = '.'): void {
    this.preSynthesize(outdir);

    // delete all generated files before we start synthesizing new ones
    cleanup(outdir);

    const constructs = ConstructNode.of(this).findAll().filter(x => x instanceof ProjectConstruct) as ProjectConstruct[];

    for (const comp of this.components) {
      comp.synthesize(outdir);
    }

    for (const child of constructs) {
      child.synthesize(outdir);
    }

    for (const comp of this.components) {
      comp.postSynthesize(outdir);
    }

    for (const child of constructs) {
      child.postSynthesize(outdir);
    }

    // project-level hook
    this.postSynthesize(outdir);


    logging.info('Synthesis complete');

    const start = this.components.find(c => c instanceof Start);
    if (start) {
      console.error();
      console.error('-'.repeat(100));
      printStartMenu();
    }

    if (this.tips.length) {
      console.error(chalk.cyanBright.underline('Tips:'));
      for (const tip of this.tips) {
        console.error(`ðŸ’¡ ${tip}`);
      }
    }
  }

  public preSynthesize(_outdir: string) {}

  /**
   * Called after synthesis. Order is *not* guaranteed.
   * @param _outdir The project directory
   */
  public postSynthesize(_outdir: string) {}

  /**
   * Adds a component to the project.
   * @internal
   */
  public _addComponent(component: Component) {
    this.components.push(component);
  }

  /**
   * @internal
   */
  public get _components(): Component[] {
    return [...this._components];
  }
}
