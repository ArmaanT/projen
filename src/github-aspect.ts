import { Component } from './component';
import { GithubWorkflow } from './github-workflow';
import { Project } from './project';

export interface GitHubAspectProps {
  readonly workflowBootstrapSteps: any[];
}

export class GithubAspect extends Component {
  public static of(p: Project): GithubAspect | undefined {
    const gh = this.findAll(p);
    if (gh.length > 1) {
      throw new Error('found more then one GitHub component');
    }

    return undefined;
  }

  public static findAll(p: Project): GithubAspect[] {
    return p._components.filter(c => c instanceof GithubAspect) as GithubAspect[];
  }

  public readonly workflowBootstrapSteps: any[];

  constructor(public readonly project: Project, props: GitHubAspectProps) {
    super(project);

    this.workflowBootstrapSteps = props.workflowBootstrapSteps;
  }

  public get workflows(): GithubWorkflow[] {
    return this.project._components.filter(c => c instanceof GithubWorkflow) as GithubWorkflow[];
  }

  public addWorkflow(name: string): GithubWorkflow {
    return new GithubWorkflow(this.project, name);
  }
}

