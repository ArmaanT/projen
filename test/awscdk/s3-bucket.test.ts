import * as s3 from '@aws-cdk/aws-s3';
import { Construct } from '@aws-cdk/core';
import { Project, AwsEnvironment } from '../../src';
import { synthSnapshot } from '../util';

test('bar', () => {
  class SecureBucket extends s3.Bucket {
    constructor(scope: Construct, id: string) {
      super(scope, id, {
        encryption: s3.BucketEncryption.KMS,
      });
    }
  }

  const project = new Project('test');

  const aws = new AwsEnvironment(project, {
    env: 'aws://11223344555/us-east-1',
  });

  new SecureBucket(aws.stack, 'SecureBucket');
  new s3.Bucket(aws.stack, 'Bucket');

  expect(synthSnapshot(project)).toMatchSnapshot();
});