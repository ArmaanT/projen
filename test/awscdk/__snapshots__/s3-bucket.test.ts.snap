// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bar 1`] = `
Object {
  "..awscdk/11223344555/us-east-1/cdk.out": "{\\"version\\":\\"6.0.0\\"}",
  "..awscdk/11223344555/us-east-1/manifest.json": Object {
    "artifacts": Object {
      "Tree": Object {
        "properties": Object {
          "file": "tree.json",
        },
        "type": "cdk:tree",
      },
      "test": Object {
        "environment": "aws://11223344555/us-east-1",
        "metadata": Object {
          "/test/Bucket/Resource": Array [
            Object {
              "data": "Bucket83908E77",
              "type": "aws:cdk:logicalId",
            },
          ],
          "/test/SecureBucket/Key/Resource": Array [
            Object {
              "data": "SecureBucketKey9F35052D",
              "type": "aws:cdk:logicalId",
            },
          ],
          "/test/SecureBucket/Resource": Array [
            Object {
              "data": "SecureBucket1ED1C5CE",
              "type": "aws:cdk:logicalId",
            },
          ],
        },
        "properties": Object {
          "templateFile": "test.template.json",
        },
        "type": "aws:cloudformation:stack",
      },
    },
    "version": "6.0.0",
  },
  "..awscdk/11223344555/us-east-1/test.template.json": Object {
    "Resources": Object {
      "Bucket83908E77": Object {
        "DeletionPolicy": "Retain",
        "Type": "AWS::S3::Bucket",
        "UpdateReplacePolicy": "Retain",
      },
      "SecureBucket1ED1C5CE": Object {
        "DeletionPolicy": "Retain",
        "Properties": Object {
          "BucketEncryption": Object {
            "ServerSideEncryptionConfiguration": Array [
              Object {
                "ServerSideEncryptionByDefault": Object {
                  "KMSMasterKeyID": Object {
                    "Fn::GetAtt": Array [
                      "SecureBucketKey9F35052D",
                      "Arn",
                    ],
                  },
                  "SSEAlgorithm": "aws:kms",
                },
              },
            ],
          },
        },
        "Type": "AWS::S3::Bucket",
        "UpdateReplacePolicy": "Retain",
      },
      "SecureBucketKey9F35052D": Object {
        "DeletionPolicy": "Retain",
        "Properties": Object {
          "Description": "Created by test/SecureBucket",
          "KeyPolicy": Object {
            "Statement": Array [
              Object {
                "Action": Array [
                  "kms:Create*",
                  "kms:Describe*",
                  "kms:Enable*",
                  "kms:List*",
                  "kms:Put*",
                  "kms:Update*",
                  "kms:Revoke*",
                  "kms:Disable*",
                  "kms:Get*",
                  "kms:Delete*",
                  "kms:ScheduleKeyDeletion",
                  "kms:CancelKeyDeletion",
                  "kms:GenerateDataKey",
                  "kms:TagResource",
                  "kms:UntagResource",
                ],
                "Effect": "Allow",
                "Principal": Object {
                  "AWS": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:",
                        Object {
                          "Ref": "AWS::Partition",
                        },
                        ":iam::11223344555:root",
                      ],
                    ],
                  },
                },
                "Resource": "*",
              },
            ],
            "Version": "2012-10-17",
          },
        },
        "Type": "AWS::KMS::Key",
        "UpdateReplacePolicy": "Retain",
      },
    },
  },
  "..awscdk/11223344555/us-east-1/tree.json": Object {
    "tree": Object {
      "children": Object {
        "Tree": Object {
          "id": "Tree",
          "path": "Tree",
        },
        "test": Object {
          "children": Object {
            "Bucket": Object {
              "children": Object {
                "Notifications": Object {
                  "id": "Notifications",
                  "path": "test/Bucket/Notifications",
                },
                "Resource": Object {
                  "attributes": Object {
                    "aws:cdk:cloudformation:props": Object {},
                    "aws:cdk:cloudformation:type": "AWS::S3::Bucket",
                  },
                  "id": "Resource",
                  "path": "test/Bucket/Resource",
                },
              },
              "id": "Bucket",
              "path": "test/Bucket",
            },
            "SecureBucket": Object {
              "children": Object {
                "Key": Object {
                  "children": Object {
                    "Resource": Object {
                      "attributes": Object {
                        "aws:cdk:cloudformation:props": Object {
                          "description": "Created by test/SecureBucket",
                          "keyPolicy": Object {
                            "Statement": Array [
                              Object {
                                "Action": Array [
                                  "kms:Create*",
                                  "kms:Describe*",
                                  "kms:Enable*",
                                  "kms:List*",
                                  "kms:Put*",
                                  "kms:Update*",
                                  "kms:Revoke*",
                                  "kms:Disable*",
                                  "kms:Get*",
                                  "kms:Delete*",
                                  "kms:ScheduleKeyDeletion",
                                  "kms:CancelKeyDeletion",
                                  "kms:GenerateDataKey",
                                  "kms:TagResource",
                                  "kms:UntagResource",
                                ],
                                "Effect": "Allow",
                                "Principal": Object {
                                  "AWS": Object {
                                    "Fn::Join": Array [
                                      "",
                                      Array [
                                        "arn:",
                                        Object {
                                          "Ref": "AWS::Partition",
                                        },
                                        ":iam::11223344555:root",
                                      ],
                                    ],
                                  },
                                },
                                "Resource": "*",
                              },
                            ],
                            "Version": "2012-10-17",
                          },
                        },
                        "aws:cdk:cloudformation:type": "AWS::KMS::Key",
                      },
                      "id": "Resource",
                      "path": "test/SecureBucket/Key/Resource",
                    },
                  },
                  "id": "Key",
                  "path": "test/SecureBucket/Key",
                },
                "Notifications": Object {
                  "id": "Notifications",
                  "path": "test/SecureBucket/Notifications",
                },
                "Resource": Object {
                  "attributes": Object {
                    "aws:cdk:cloudformation:props": Object {
                      "bucketEncryption": Object {
                        "serverSideEncryptionConfiguration": Array [
                          Object {
                            "serverSideEncryptionByDefault": Object {
                              "kmsMasterKeyId": Object {
                                "Fn::GetAtt": Array [
                                  "SecureBucketKey9F35052D",
                                  "Arn",
                                ],
                              },
                              "sseAlgorithm": "aws:kms",
                            },
                          },
                        ],
                      },
                    },
                    "aws:cdk:cloudformation:type": "AWS::S3::Bucket",
                  },
                  "id": "Resource",
                  "path": "test/SecureBucket/Resource",
                },
              },
              "id": "SecureBucket",
              "path": "test/SecureBucket",
            },
          },
          "id": "test",
          "path": "test",
        },
      },
      "id": "App",
      "path": "",
    },
    "version": "tree-0.1",
  },
}
`;
